### ISSUE TITLE
Canonical way to get normal during surface assembly

### ISSUE THREAD

**Issue by szellmann** [szellmann/visionaray#3]
Provide a canonical way to obtain a surface (or vertex) normal when assembling surfaces. This would enable real user-defined geometry.

**Comment by szellmann** [comment]
Partially addressed by commit 21cc07a21878ea9880c88f8e2466bb5424c70c22. Some primitives have precalculated normals, some don't. `get_surface()` should probably be adapted so that the following versions of the `get_normal()` customization point are valid:
```
get_normal(hit_rec, prim);
get_normal(normlas, hit_rec, Primitive{});
get_normal(normals, hit_rec, Primitive{}, NormalBinding{});
```
Similar considerations apply to `get_tex_coord()`.

**Comment by szellmann** [comment]
63c20cfbd4ec9c8f47643c2c07da1b546558f048 and address adf5861d2b2def84baab7d5a16d17a1761998e5c this. I'm keeping the issue open because we should also get rid of the NormalBinding parameter in cases where this does not apply.

**Comment by szellmann** [comment]
Solved via cebec90b2ef6cacd78a55c525cb2e7a8a364ce39

Interface is now:

```
get_normal(hit_rec, prim);
get_normal(normals, hit_rec, Primitive{}, NormalBinding{});
```

i.e. with precalculated normals around, specifying the normal binding is mandatory.
